implementations/rust/ockam/ockam_api/src/authenticator/direct.rs

```diff
@@ -0,0 +1,419 @@
+use core::time::Duration;
+use core::{fmt, str};
+use minicbor::encode::write::Cursor;
+use minicbor::{Decoder, Encode};
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::{self, Address, Result, Route, Routed, Worker};
+use ockam_identity::authenticated_storage::AuthenticatedStorage;
+use ockam_identity::{IdentityIdentifier, IdentitySecureChannelLocalInfo};
+use ockam_node::Context;
+use tracing::{trace, warn};
+use types::{AddEnroller, AddMember, Placeholder};
+
+// storage scopes:
+const ENROLLER: &str = "enroller";
+const MEMBER: &str = "member";
+
+const MAX_VALIDITY: Duration = Duration::from_secs(2 * 3600);
+
+#[derive(Debug)]
+pub struct Server<S, T> {
+    m_store: S, // member store
+    e_store: T, // enroller store
+    signer: signer::Client,
+    admin: Option<IdentityId<'static>>,
+}
+
+#[ockam_core::worker]
+impl<S, T> Worker for Server<S, T>
+where
+    S: AuthenticatedStorage,
+    T: AuthenticatedStorage,
+{
+    type Context = Context;
+    type Message = Vec<u8>;
+
+    async fn handle_message(&mut self, c: &mut Context, m: Routed<Self::Message>) -> Result<()> {
+        if let Ok(i) = IdentitySecureChannelLocalInfo::find_info(m.local_message()) {
+            let r = self.on_request(i.their_identity_id(), m.as_body()).await?;
+            c.send(m.return_route(), r).await
+        } else {
+            let mut dec = Decoder::new(m.as_body());
+            let req: Request = dec.decode()?;
+            let res = response::forbidden(&req, "secure channel required").to_vec()?;
+            c.send(m.return_route(), res).await
+        }
+    }
+}
+
+impl<S, T> Server<S, T>
+where
+    S: AuthenticatedStorage,
+    T: AuthenticatedStorage,
+{
+    pub fn new(m_store: S, e_store: T, signer: signer::Client) -> Self {
+        Server {
+            m_store,
+            e_store,
+            signer,
+            admin: None,
+        }
+    }
+
+    pub fn set_admin(&mut self, id: &IdentityIdentifier) {
+        self.admin = Some(IdentityId::new(id.key_id().to_string()))
+    }
+
+    async fn on_request(&mut self, from: &IdentityIdentifier, data: &[u8]) -> Result<Vec<u8>> {
+        let mut dec = Decoder::new(data);
+        let req: Request = dec.decode()?;
+
+        trace! {
+            target: "ockam_api::authenticator::direct::server",
+            from   = %from,
+            id     = %req.id(),
+            method = ?req.method(),
+            path   = %req.path(),
+            body   = %req.has_body(),
+            "request"
+        }
+
+        let res = match req.method() {
+            Some(Method::Post) => match req.path_segments::<2>().as_slice() {
+                // Admin wants to add an enroller.
+                ["enroller"] if self.is_admin(from) => {
+                    let add: AddEnroller = dec.decode()?;
+                    let enroller = minicbor::to_vec(Placeholder)?;
+                    self.e_store
+                        .set(ENROLLER, add.enroller().as_str().to_string(), enroller)
```

Comment:
Maybe it could look something like this?
```
self.e_store.set(add.enroller_id()), "is_enroller", true)
```

Reply:
The authenticated storage layout was not really specified and is currently not restricted to entries of the type `IdentityId × String × Bytes`. Originally the thought was that the `Bytes` values would be signed data to allow validating the authenticity again but this may have been overtaken by events. The first element is merely a scope to hold key-value pairs and a means to avoid collisions when multiple services share a single storage. In the case of globally unique entries such as identity IDs we can change that and store them the way you proposed but I am not sure if `AuthenticatedStorage` should be made specific for identities.

Reply:
Unless we have use cases where id is not and IdentityIdentifier, let's use it this way. Other use-cases could probably have different traits

Reply:
https://github.com/build-trust/ockam/commit/5967700e0150bafef2b26dc14c3b48971ed6246f